### DISCUSSION TITLE
Implement the ability to partially compare tuples

### DISCUSSION THREAD

**Discussion by ghost** [dotnet/csharplang#7070] [Language Ideas]
Some thoughts on this feature.

**Example 1:**

from
```c#
(int left, int right) value = (0, 1);

if(value.left is 0)
{
   //
}
```
to
```c#
(int left, int right) value = (0, 1);

if(value is (0, _))
{
   //
}
```

**Example 2:**

from
```c#
(Node left, Node mid, Node right) value = (new Node(), new Node(), new Node());

if(value.left is not null && value.right is null)
{
   //
}
```
to
```c#
(Node left, Node mid, Node right) value = (new Node(), new Node(), new Node());

if(value is (not null, _, null)) // if(value is (!, _, ?))
{
   //
}
```

**Some notations:**

for null value - null, ? -> (null, null), (?, ?)
for not null value - not null, ! -> (not null, not null), (!, !)
for any value - any, _ -> (any, any), (_, _)

I think in some cases it will simplify the code and improve readability. What do you think?

**Comment by HaloFour** [comment]
The examples that you're suggesting already work, positional patterns permit recursive subpatterns: https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/patterns#positional-pattern

As for the shorthand notation, I personally don't think that it makes the code clearer.

**Comment by DavidArno** [comment]
There are already three ways of testing for "not null":
```cs
(Node? left, Node? mid, Node? right) value = (new Node(), new Node(), null);

if(value is (not null, _, null)) 
{
   Console.WriteLine("matches");
}

if(value is ({}, _, null)) 
{
   Console.WriteLine("matches");
}

if(value is (Node, _, null)) 
{
   Console.WriteLine("matches");
}
```
I really don't see how adding `!` as a fourth way brings any value to the language. I'd be strongly opposed to this as it just complicates things in my view.

As for `?`, when used as the "Elvis operator" (`?.`), it effectively represents "not null". With the ternary operator (`?:`), it effectively represents `true` and when used to denote a nullable type (`T?`) is effectively means "can be null". Using it here to mean "definitely null" seems confusing. `??` more means `null` in my view, but again is there any real benefit to having `??` be an alternative to `null` in pattern matching? I think the extra complexity of the language this creates far far outweighs the benefit of typing two characters.
